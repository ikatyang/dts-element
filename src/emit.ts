import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
