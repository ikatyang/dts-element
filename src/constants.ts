import * as ts from 'typescript';
import { create_native_type, INativeType } from './types/native-type';

export enum ElementKind {
  ArrayType,
  ClassDeclaration,
  ClassMember,
  ConditionalType,
  ConstructorType,
  EnumDeclaration,
  ExportDefault,
  ExportEqual,
  ExportMember,
  ExportNamed,
  ExportNamespace,
  FunctionDeclaration,
  FunctionType,
  GeneralType,
  GenericDeclaration,
  GlobalDeclaration,
  ImportEqual,
  ImportMember,
  ImportNamed,
  ImportNamespace,
  IndexSignature,
  InferType,
  InterfaceDeclaration,
  IntersectionType,
  JSDocComment, // tslint:disable-line:naming-convention
  KeyofType,
  LiteralType,
  MappedType,
  ModuleDeclaration,
  MultiLineComment,
  NamespaceDeclaration,
  NativeType,
  ObjectMember,
  ObjectType,
  ParameterDeclaration,
  SingleLineComment,
  SubType,
  TopLevelElement,
  TripleSlashReference,
  TupleType,
  TypeDeclaration,
  TypePredicate,
  TypeofType,
  UnionType,
  VariableDeclaration,
}

export const any_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
});
export const boolean_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword),
});
export const never_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword),
});
export const null_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword),
});
export const number_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
});
export const object_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
});
export const string_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
});
export const symbol_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword),
});
export const this_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword),
});
export const undefined_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword),
});
export const void_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword),
});
