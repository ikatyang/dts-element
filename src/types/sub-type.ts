import * as ts from 'typescript';
import { IType } from '../collections';
import { any_type, ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';

export interface ISubTypeOptions extends IElementOptions {
  object: IType;
  index?: IType;
}

export interface ISubType
  extends IElement<ElementKind.SubType>,
    ISubTypeOptions {}

export const create_sub_type = (options: ISubTypeOptions): ISubType => ({
  ...create_element(ElementKind.SubType),
  ...options,
});

export const is_sub_type = (value: any): value is ISubTypeOptions =>
  is_element(value) && value.kind === ElementKind.SubType;

/**
 * @hidden
 */
export const transform_sub_type = (element: ISubType, path: IElement<any>[]) =>
  ts.createIndexedAccessTypeNode(
    transform(element.object, path) as ts.TypeNode,
    transform(element.index || any_type, path) as ts.TypeNode,
  );
