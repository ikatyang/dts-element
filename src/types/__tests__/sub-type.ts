import { any_type, number_type, string_type } from '../../constants';
import { emit } from '../../emit';
import { create_sub_type, is_sub_type } from '../sub-type';

it('should return correctly with object', () => {
  expect(
    emit(
      create_sub_type({
        object: string_type,
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with object and index', () => {
  expect(
    emit(
      create_sub_type({
        object: create_sub_type({
          object: any_type,
          index: string_type,
        }),
        index: number_type,
      }),
    ),
  ).toMatchSnapshot();
});

describe('is_sub_type', () => {
  it('should return correctly', () => {
    const element = create_sub_type({
      object: create_sub_type({
        object: any_type,
        index: string_type,
      }),
      index: number_type,
    });
    expect(is_sub_type(element)).toBe(true);
  });
});
