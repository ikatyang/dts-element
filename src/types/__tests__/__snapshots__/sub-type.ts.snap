// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with object 1`] = `"string[any]"`;

exports[`should return correctly with object and index 1`] = `"any[string][number]"`;
